File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/1f6c549a9d407c6276f7b0cacba3bb49.java
import java.util.*;
import java.io.*;

class Main {
    static final int MOD = 998244353;
    static final int K = 18;

    static class Mint {
        int n;

        Mint(int n) {
            this.n = n % MOD;
            if (this.n < 0) this.n += MOD;
        }

        Mint() {
            this(0);
        }

        Mint inv() {
            return pow(MOD - 2);
        }

        Mint pow(int exp) {
            Mint result = new Mint(1);
            Mint base = new Mint(n);
            while (exp > 0) {
                if ((exp & 1) == 1) result = result.mul(base);
                base = base.mul(base);
                exp >>= 1;
            }
            return result;
        }

        Mint mul(Mint other) {
            return new Mint((int) ((long) n * other.n % MOD));
        }

        Mint add(Mint other) {
            return new Mint(n + other.n);
        }

        Mint sub(Mint other) {
            return new Mint(n - other.n);
        }

        @Override
        public String toString() {
            return String.valueOf(n);
        }
    }

    static class Poly {
        Mint[] c;

        Poly(int n) {
            c = new Mint[n + 1];
            Arrays.fill(c, new Mint());
        }

        int deg() {
            return c.length - 1;
        }

        Poly log() {
            assert c[0].n == 1;
            int n = deg();
            Poly g = new Poly(n);
            for (int i = 0; i < n; i++) {
                g.c[i] = new Mint(i + 1).mul(c[i + 1]);
            }
            return g;
        }

        Poly exp() {
            assert c[0].n == 0;
            int n = deg();
            Poly result = new Poly(n);
            result.c[0] = new Mint(1);
            for (int i = 1; i <= n; i++) {
                for (int j = 0; j < i; j++) {
                    result.c[i] = result.c[i].add(new Mint(c[j].n).mul(result.c[i - j - 1]));
                }
            }
            return result;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int k = scanner.nextInt();
        int[] hs = new int[n];
        for (int i = 0; i < n; i++) {
            hs[i] = scanner.nextInt();
        }

        int cnt = 0;
        for (int i = 0; i < n; i++) {
            if (hs[i] == hs[(i + 1) % n]) cnt++;
        }

        if (cnt == n) {
            System.out.println(0);
            return;
        }

        Poly A = new Poly(200000 + 1000);
        A.c[0] = new Mint(1);
        A.c[1] = new Mint((k - 2) % MOD);
        A.c[2] = new Mint(1);

        Poly B = A.log();
        for (int i = 0; i < B.c.length; i++) {
            B.c[i] = new Mint(B.c[i].n * (n - cnt) % MOD);
        }

        Poly C = B.exp();

        Mint ans = new Mint((int) Math.pow(k % MOD, n - cnt));
        for (int i = 0; i <= n - cnt; i++) {
            ans = ans.sub(C.c[i]);
        }

        ans = ans.mul(new Mint((int) Math.pow(k % MOD, cnt)));
        System.out.println(ans);
    }
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Python/1f6c549a9d407c6276f7b0cacba3bb49.py
mod = 998244353
K = 18

class FastIO:
    def __init__(self):
        pass

n, k = map(int, input().split())
hs = list(map(int, input().split()))

cnt = 0
for i in range(n):
    if hs[i] == hs[(i+1)%n]:
        cnt += 1

if cnt == n:
    print(0)
    exit()

class Poly:
    def __init__(self, n):
        self.c = [0] * (n + 1)

A = Poly(2e5+1000)
A.c[0] = 1
A.c[1] = (k-2) % mod
A.c[2] = 1

def power(x, n):
    res = 1
    while n > 0:
        if n % 2 == 1:
            res = (res * x) % mod
        x = (x * x) % mod
        n //= 2
    return res

def inv(x):
    return power(x, mod-2)

def add(x, y):
    return (x + y) % mod

def sub(x, y):
    return (x - y + mod) % mod

def mul(x, y):
    return (x * y) % mod

def div(x, y):
    return (x * inv(y)) % mod

def log(p):
    n = len(p.c)
    res = Poly(n)
    for i in range(1, n):
        res.c[i-1] = mul(p.c[i], inv(i))
    return res

def exp(p):
    n = len(p.c)
    res = Poly(n)
    res.c[0] = 1
    for i in range(1, n):
        for j in range(1, i+1):
            res.c[i] = add(res.c[i], mul(p.c[j], res.c[i-j]))
        res.c[i] = div(res.c[i], i)
    return res

B = log(A)
for i in range(len(B.c)):
    B.c[i] = mul(B.c[i], n-cnt)

C = exp(B)

ans = power(k, n-cnt)
for i in range(len(C.c)):
    ans = sub(ans, C.c[i])

ans = mul(ans, power(k, cnt))
print(ans)
